server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperation.java

```diff
@@ -68,24 +67,36 @@ public DelayedOperation(long delayMs, Lock lock) {
      * Return true iff the operation is completed by the caller: note that
      * concurrent threads can try to complete the same operation, but only
      * the first thread will succeed in completing the operation and return
-     * true, others will still return false
+     * true, others will still return false.
      */
     public boolean forceComplete() {
-        if (completed.compareAndSet(false, true)) {
-            // cancel the timeout timer
-            cancel();
-            onComplete();
-            return true;
-        } else {
+        // Do not proceed if the operation is already completed.
+        if (completed) {
             return false;
         }
+        // Attain lock prior completing the request.
+        lock.lock();
```

Comment:
After this change, the lock must be reentrant, correct? If so, we should consider replacing the `Lock` type with `ReentrantLock` to prevent potential deadlocks.

Additionally, `DelayedOperation(long delayMs, Optional<Lock> lockOpt)` seems to lose its meaning once the old group coordinator is removed. We might want to clean that up.

These points can be addressed in a follow-up.

Reply:
> Are you suggesting to make it more tighter and use ReentrantLock explicitly?

yes, but this change needs to cleanup the code. It can be addressed later if we want to merge this bug fix first.

Reply:
Sure, I ll create a folow up.

Reply:
I was a little bored, so I created Jira KAFKA-19322 (https://issues.apache.org/jira/browse/KAFKA-19322). Feel free to take it over, or I can assign it to a mentee :smile: